src/Microsoft.OData.Core/PropertyCache.cs

```diff
@@ -14,15 +15,68 @@ namespace Microsoft.OData
     /// </summary>
     internal class PropertyCache
     {
-        private readonly Dictionary<string, PropertySerializationInfo> propertyInfoDictionary = new Dictionary<string, PropertySerializationInfo>();
+        private readonly Dictionary<PropertyKey, PropertySerializationInfo> propertyInfoDictionary = new Dictionary<PropertyKey, PropertySerializationInfo>();
 
-        public PropertySerializationInfo GetProperty(IEdmModel model, string name, string uniqueName, IEdmStructuredType owningType)
+        private struct PropertyKey : IEquatable<PropertyKey>
+        {
+            public PropertyKey(int depth, string name, string fullTypeName)
+            {
+                this.Depth = depth;
+                this.Name = name;
+                this.FullTypeName = fullTypeName;
+            }
+
+            public int Depth { get; private set; }
+
+            public string Name { get; private set; }
+         
+            public string FullTypeName { get; private set; }
+
+            public static bool operator ==(PropertyKey lhs, PropertyKey rhs)
+            {
+                return lhs.Equals(rhs);
+            }
+
+            public static bool operator !=(PropertyKey lhs, PropertyKey rhs)
+            {
+                return !lhs.Equals(rhs);
+            }
+
+            public override int GetHashCode()
+            {
+                return this.Depth.GetHashCode() ^
+                       this.Name.GetHashCode() ^
+                       (this.FullTypeName?.GetHashCode() ?? 0);
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (obj is PropertyKey propertyKey)
+                {
+                    return this.Equals(propertyKey);
+                }
+
+                return false;
+            }
+
+            public bool Equals(PropertyKey other)
+            {
+                return
+                    this.Depth == other.Depth &&
+                    this.Name == other.Name &&

```

Comment:
Should this use String.Equals(this.Name, other.Name, StringComparison.InvariantCulture) ?

Reply:
I tried to preserve the old behavior which is Ordinal. But let me know if this should be Invariant. Though I rather have Ordinal for performance :)